unittest/model-graph.cpp

```diff
@@ -0,0 +1,189 @@
+#include "pinocchio/multibody/joint/joint-free-flyer.hpp"
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(ModelGraph)
+
+BOOST_AUTO_TEST_CASE(test_add_vertex)
+{
+  pinocchio::ModelGraph g;
+
+  g.addBody("body1", pinocchio::Inertia::Identity());
+  BOOST_CHECK(g.name_to_vertex.find("body1") != g.name_to_vertex.end());
+}
+
+BOOST_AUTO_TEST_CASE(test_add_edges)
+{
+  typedef boost::adjacency_list<
+    boost::vecS, boost::vecS, boost::directedS, pinocchio::ModelGraphVertex,
+    pinocchio::ModelGraphEdge>
+    Graph;
+  typedef typename boost::graph_traits<Graph>::vertex_descriptor VertexDesc;
+
+  pinocchio::ModelGraph g;
+  ////////////////////////////////////// Bodies
+  g.addBody("body1", pinocchio::Inertia::Identity());
+  g.addBody("body2", pinocchio::Inertia::Identity());
+
+  /////////////////////////////////////// Joints
+  g.addJoint(
+    "body1_to_body2",
+    pinocchio::internal::JointRevoluteGraph("body1_to_body2", Eigen::Vector3d::UnitZ()), "body1",
+    pinocchio::SE3(Eigen::Matrix3d::Identity(), Eigen::Vector3d(2., 0., 0.)), "body2",
+    pinocchio::SE3(Eigen::Matrix3d::Identity(), Eigen::Vector3d(2., 2., 0.)));
+
+  VertexDesc v_out = g.name_to_vertex["body1"];
+  VertexDesc v_in = g.name_to_vertex["body2"];
+  BOOST_CHECK(boost::edge(v_out, v_in, g.g).second);
+  BOOST_CHECK(boost::edge(v_in, v_out, g.g).second);
+}
+
+BOOST_AUTO_TEST_CASE(test_linear_robot)
+{
+  pinocchio::ModelGraph g;
+  //////////////////////////////////////// Bodies
+  g.addBody("body1", pinocchio::Inertia::Identity());
+  g.addBody(
+    "body2",
+    pinocchio::Inertia(4., pinocchio::Inertia::Vector3(0., 2., 0.), pinocchio::Symmetric3::Zero()));
+  g.addBody(
+    "body3", pinocchio::Inertia(
+               3., pinocchio::Inertia::Vector3(0., -3, -2.), pinocchio::Symmetric3::Zero()));
+
+  /////////////////////////////////////// Joints
+  g.addJoint(
+    "body1_to_body2",
+    pinocchio::internal::JointRevoluteGraph("body1_to_body2", Eigen::Vector3d::UnitZ()), "body1",
+    pinocchio::SE3(Eigen::Matrix3d::Identity(), Eigen::Vector3d(2., 0., 0.)), "body2",
+    pinocchio::SE3(Eigen::Matrix3d::Identity(), Eigen::Vector3d(2., 2., 0.)));
+  g.addJoint(
+    "body2_to_body3",
+    pinocchio::internal::JointRevoluteGraph("body2_to_body3", Eigen::Vector3d::UnitX()), "body2",
+    pinocchio::SE3(Eigen::Matrix3d::Identity(), Eigen::Vector3d(2., -3., 0.)), "body3",
+    pinocchio::SE3(Eigen::Matrix3d::Identity(), Eigen::Vector3d(4., -3., 0.)));
+
+  ///////////////// Model
+  pinocchio::Model m = g.buildModel("body1", pinocchio::JointModelFreeFlyer());
+
+  BOOST_CHECK(m.jointPlacements[m.getJointId("body1_to_body2")].isApprox(
+    pinocchio::SE3(Eigen::Matrix3d::Identity(), Eigen::Vector3d(2, 0, 0))));
+  BOOST_CHECK(m.jointPlacements[m.getJointId("body2_to_body3")].isApprox(
+    pinocchio::SE3(Eigen::Matrix3d::Identity(), Eigen::Vector3d(4, -1, 0))));
+
+  pinocchio::Model m1 = g.buildModel("body3", pinocchio::JointModelFreeFlyer());
```

Comment:
We should check the jointPlacement also:

- Add comments to explains how jointPlacements are manually computed
- Use FK with manually done equivalent Q vector to test the equal X_0_i (with a m.X_0_i[b3] transformation)

A